### ISSUE TITLE
Cluster API deletion lifecycle improvement

### ISSUE THREAD

**Issue by chuckha** [kubernetes-sigs/cluster-api#1779]
> ⚠️ Cluster API maintainers can ask to turn an issue-proposal into a [CAEP](https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/YYYYMMDD-template.md) when necessary, this is to be expected for large changes that impact multiple components, breaking changes, or new large features.

***Goals***
1. Clean up workload clusters before deleting them which allows controllers running on the workload cluster to be responsible for their object's clean up lifecycle phase.
1. Ensure opt-out is possible/force delete in case the cluster is in a bad state and cannot handle deletes (I am still a maybe on opt-out because this could, like today's behavior, orphan resources).
1. Keep the responsibility of deletion of cloud-resources CAPI did not create out of CAPI & providers.

***Non-Goals/Future Work***
1. Clean up after a cluster that is in a very broken state where a cluster can no longer accept delete requests. This is the responsibility of the owner to clean up resources they provisioned.

**User Story**

As a developer I would like my cloud-provider created cloud resources (LBs are the most common, but others such as PVs exist) to be cleaned up when I delete my cluster. 

As an operator I would like deleting a cluster to call resource deletion on every resource in my cluster to ensure the clean up lifecycle phase can happen on any resource where it is required.

**Detailed Description**

For a concrete example let's take a look at Kubernetes Services of Type: LoadBalancer. The implication here is that the Kubernetes cluster is running a cloud-provider that knows what to do when they see Services created with this type. The cloud-provider will create a LoadBalancer and tag it with a special cloud-provider tag. We have [special code in CAPA][special] to delete these cloud-provider provisioned resources. However, this code is already handled in the aws cloud-provider. CAPA did not create these load balancers. We just happen to know that they block deletion sometimes so we've added special code. We are by-passing the clean-up phase of delete by trying to implement it ourselves. What about PVs? We don't clean those up because those don't block the deletion of a Cluster, but they should be cleaned up just like LoadBalancers are cleaned up (if not asked to be retained).

[special]: https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/1168

I propose that we centralize delete behavior in CAPI to delete *all* resources in a workload cluster before actually tearing down the CAPI Cluster infrastructure. This would be a major change to the cluster-controller. This would require us to figure out all the types that are registered on the workload cluster, list every resource the workload cluster has and issue a delete for every namespaced object outside of the kube-system namespace (exact details TBD). This allows the deletions of external resources (such as cloud infrastructure) to be removed via the controller that created them in the first place, such as the cloud-provider controller.

**Contract changes [optional]**
No contract changes.

**Data model changes [optional]**
No data model changes

**Downsides**
This could increase the deletion of the cluster object in proportion with the amount of cleanup/resources that exist in the workload cluster. 

/kind proposal

**Comment by michaelgugino** [comment]
I think the most important downside is in the case of a broken cluster.  If the cluster is not operational or is otherwise unable to remove those resources, cluster-api will be blocked on deleting the cluster.

**Comment by ncdc** [comment]
Yes, this is yet another tricky situation when it comes to automating cluster lifecycle management. What happens if things are stuck - do we try to delete a few times, and then force delete?

Another possible gotcha includes deleting aggregated api servers. If you delete the pod for an aggregated api server and then you later try to perform any CRUD operations on that server's resources, those operations will fail. Order definitely matters.

**Comment by vincepri** [comment]
As per April 27th 2020 community guidelines, this project follows the process outlined in https://github.com/kubernetes-sigs/cluster-api/blob/master/CONTRIBUTING.md#proposal-process-caep for large features or changes.

Following those guidelines, I'll go ahead and close this issue for now and defer to contributors interested in pushing the proposal forward to open a collaborative document proposal instead, and follow the process as described.

/close
